overlord/snapstate/snapstate.go

```diff
@@ -4146,6 +4147,125 @@ func GadgetConnections(st *state.State, deviceCtx DeviceContext) ([]gadget.Conne
+	for _, seq := range snapst.Sequence {
+		neededSnapDownloads = append(neededSnapDownloads, seq.Revision)
+	}
+	// keep revision in refresh hint (if it exists)
+	var refreshHints map[string]*refreshCandidate
+	if err := st.Get("refresh-candidates", &refreshHints); err != nil {
+		if !errors.Is(err, &state.NoStateError{}) {
+			return nil, fmt.Errorf("cannot get refresh-candidates: %v", err)
+		}
+	} else if hint, ok := refreshHints[snapName]; ok {
+		neededSnapDownloads = append(neededSnapDownloads, hint.Revision())
+	}
+	// keep revisions pointed to by a download task in an ongoing change
+	for _, chg := range st.Changes() {
+		if chg.IsReady() {
+			continue
+		}
+		for _, t := range chg.Tasks() {
+			if t.Kind() != "download-snap" {
+				continue
+			}
+			snapsup, _, err := snapSetupAndState(t)
+			if err != nil {
+				return nil, err
+			}
+			if snapsup.InstanceName() != snapName {
+				continue
+			}
+			neededSnapDownloads = append(neededSnapDownloads, snapsup.Revision())
+		}
+	}
+
+	return neededSnapDownloads, nil
+}
+
+// cleanDownloads removes downloads that are no longer needed for all snaps.
+//
+// A download is kept only if:
+//    1. Revision is in a snap's sequence
+//    2. Revision is in a snap's refresh hint
+//    3. Revision is pointed to by a download task in an ongoing change
+//
+// It is the caller's responsibility to lock state before calling this function.
+var cleanDownloads = func(st *state.State) error {
+	snapStates, err := All(st)
+	if err != nil {
+		return err
+	}
+
+	keepContent := map[string]osutil.FileState{}
+	for snapName := range snapStates {
+		neededSnapRevisions, err := deriveNeededSnapRevisions(st, snapName)
+		if err != nil {
+			return err
+		}
+		for _, revision := range neededSnapRevisions {
+			mountFile := snap.MountFile(snapName, revision)
+			if osutil.FileExists(mountFile) {
+				keepContent[filepath.Base(mountFile)] = &osutil.FileReference{Path: mountFile}
+			}
+		}
+	}
+
+	_, _, err = osutil.EnsureDirState(dirs.SnapBlobDir, "*.snap", keepContent)
+	return err
+}
+
+// cleanSnapDownloads removes downloads that are no longer needed for the given snap.
+//
+// A download is kept only if:
+//    1. Revision is in snap's sequence
+//    2. Revision is in snap's refresh hint
+//    3. Revision is pointed to by a download task in an ongoing change
+//
+// It is the caller's responsibility to lock state before calling this function.
+var cleanSnapDownloads = func(st *state.State, snapName string) error {
+	neededSnapRevisions, err := deriveNeededSnapRevisions(st, snapName)
+	if err != nil {
+		return err
+	}
+
+	keepContent := map[string]osutil.FileState{}
+	for _, revision := range neededSnapRevisions {
+		mountFile := snap.MountFile(snapName, revision)
+		if osutil.FileExists(mountFile) {
+			keepContent[filepath.Base(mountFile)] = &osutil.FileReference{Path: mountFile}
+		}
+	}
+
+	globs := []string{
+		fmt.Sprintf("%s_[0-9].snap", snapName),
+		fmt.Sprintf("%s_x[0-9].snap", snapName),
```

Comment:
These patterns won't pick up snap revision with more than 2 digits which doesn't seem right

Reply:
Great catch, Thank you!